### ISSUE TITLE
Make LocalPackageFileCache methods virtual

### ISSUE THREAD

**Issue by AndreiTimisescu** [NuGet/Home#10325]
## Details about Problem

NuGet product used (any):

NuGet version (any):

dotnet.exe --version (any):

VS version (any):

OS version (any):

Worked before? never

## Detailed repro steps so we can see the same problem

We are using NuGet.Commands.RestoreCommand to restore nuget packages and store them on Azure Storage.

Unfortunately LocalPackageFileCache is very slow and in order to optimize it we need to extend it. Unfortunately the class does not implement any interface nor has virtual methods.

Set fallback folders and install path to a disk with slow access. Restoring from it will be slow even if all packages are cached because LocalPackageFileCache try to access the actual files. Still in specific user scenario checking for the actual disk files may be avoid and performance may be improved.

## Other suggested things

Making LocalPackageFileCache methods virtual would be the minimum requirement to give NuGet clients an extension point for optimizations.

### Verbose Logs

Not applicable

### Sample Project

Not applicable

```

**Comment by nkolev92** [comment]
Thanks for creating a PR @AndreiTimisescu 

I'd like to learn more about the motivation. 

Can you be more specific about what parts are slow?
Are there any optimizations we can apply to the main APIs, so that everyone can benefit from the changes?

**Comment by AndreiTimisescu** [comment]
> be more specific about what parts are slow?
> Are there any optimizations we can apply to the main APIs, so that everyone can benefit from the changes?

Hi, we are developing a web application using Azure ecosystem. One of the backend services need to restore user projects and serve the restored dlls to be loaded in a Blazor application running on front end. The restore operation needs to be fast as it happens every time the user opens a project. In most cases the dlls will be already restored on backend and stored on Azure Storage as there will be a crawler installing packages for some predefined feeds. But file operations in Azure Storage are slow and unnecessary as the information regarding installed packages can be stored in a database and cached. Unfortunately LocalPackageFileCache is not extendable.

Just opening LocalPackageFileCache for extension will be enough to allow the NuGet API clients implement their optimized version for their specific scenarios. The one implementation fits all scenarios proved to be a  performance bottleneck in our use case.

Thank you very much for looking into this.

**Comment by zivkan** [comment]
I'm not (yet?) an expert on NuGet's restore code, but I'm concerned that there may be a fair amount of code that assumes that the files or packages returned by this class are really local on disk, available though `System.IO` APIs. If you have an implementation that queries a database or blob storage, so the files are not really local, other parts of the product might fail when it tries to use those `System.IO` APIs.

Have you tested your changes to verify that it really works as you need?

I assume that you're only restoring, not really building the project, otherwise I don't see how the .NET SDK and compilers would work. Or are you using file shares via Azure Storage, not Blob Storage? Are you using this as a package source, or as the global packages folder?

You might be better off exposing your cached packages as a NuGet HTTP package source, and then NuGet's `LocalPackageFileCache` really will use the VM's local disk only, which I expect to be fast enough. [Sleet](https://github.com/emgarten/Sleet) can be used to create a NuGet HTTP feed that can be hosted in blob storage (assuming you don't need package search), although I'm not sure how the authentication will work if the storage container is not public. Another option is [BaGet](https://github.com/loic-sharma/BaGet), which has Azure hosting built-in, with packages stored in Blob Storage. I'm not sure it has authentication built-in, but I've used this product multiple times when prototyping/testing different features, and I found it very easy to extend for whatever feature I was testing.

If you can expose your Azure Storage package cache as an HTTP NuGet feed, then you might need a lot less custom code, which would be a maintenance win for you. But I also might not fully understand what you're doing, so maybe this won't work afterall.

**Comment by AndreiTimisescu** [comment]
> I'm not (yet?) an expert on NuGet's restore code, but I'm concerned that there may be a fair amount of code that assumes that the files or packages returned by this class are really local on disk, available though `System.IO` APIs. If you have an implementation that queries a database or blob storage, so the files are not really local, other parts of the product might fail when it tries to use those `System.IO` APIs.
> 
> Have you tested your changes to verify that it really works as you need?
> 
> I assume that you're only restoring, not really building the project, otherwise I don't see how the .NET SDK and compilers would work. Or are you using file shares via Azure Storage, not Blob Storage? Are you using this as a package source, or as the global packages folder?
> 
> You might be better off exposing your cached packages as a NuGet HTTP package source, and then NuGet's `LocalPackageFileCache` really will use the VM's local disk only, which I expect to be fast enough. [Sleet](https://github.com/emgarten/Sleet) can be used to create a NuGet HTTP feed that can be hosted in blob storage (assuming you don't need package search), although I'm not sure how the authentication will work if the storage container is not public. Another option is [BaGet](https://github.com/loic-sharma/BaGet), which has Azure hosting built-in, with packages stored in Blob Storage. I'm not sure it has authentication built-in, but I've used this product multiple times when prototyping/testing different features, and I found it very easy to extend for whatever feature I was testing.
> 
> If you can expose your Azure Storage package cache as an HTTP NuGet feed, then you might need a lot less custom code, which would be a maintenance win for you. But I also might not fully understand what you're doing, so maybe this won't work afterall.

Thanks Andy, for taking time to look into this.

In our configuration, the files are accessible via System.IO as the Azure Storage is mounted as a local drive, the problem is that the access is slow. The packages are restored in a temporary packages folder and then moved on feed related fallback folders stored in Azure Storage. Users have governed access to different feeds and their corresponding related fallback folders. 

I already have a version of nuget packages compiled with the changes I proposed and the custom implementation works - restoring (calculating dependencies) for log4net and serving them to the blazor front end application, when the packages were  already restored takes 6 seconds without the optimization and 400ms with the optimization. 

The problem is that it would be grate if we don't have to compile NuGet and publish some custom nuget version which we have to maintain. 

Long story short: access to local packages stored on a mounted cloud drive is slow, and an extension point for LocalPackageFileCache (interface or virtual methods) is required to let users optimize it.
